src/lib/HoneylemonService.js

```diff
@@ -0,0 +1,380 @@
+
+    return {
+      price,
+      resultOrders,
+      ordersRemainingFillableTakerAssetAmounts,
+      takerAssetFillAmounts,
+      remainingFillAmount
+    };
+  }
+
+  async get0xFeeForOrderBatch(gasPrice, batchSize) {
+    const protocolFeeMultiplier = await this.contractWrappers.exchange
+      .protocolFeeMultiplier()
+      .callAsync();
+    return new BigNumber(protocolFeeMultiplier)
+      .multipliedBy(gasPrice)
+      .multipliedBy(batchSize);
+  }
+
+  async estimateGas(signedOrders, takerAssetFillAmounts, takerAddress) {
+    const signatures = signedOrders.map(o => o.signature);
+    const gasPrice = 10e9; // Set to 10GWEI
+    const value = await this.get0xFeeForOrderBatch(gasPrice, signedOrders.length);
+
+    const gas = await this.contractWrappers.exchange
+      .batchFillOrKillOrders(signedOrders, takerAssetFillAmounts, signatures)
+      .estimateGasAsync({ from: takerAddress, value, gasPrice });
+
+    return gas;
+  }
+
+  getFillOrdersTx(signedOrders, takerAssetFillAmounts) {
+    const signatures = signedOrders.map(o => o.signature);
+    return this.contractWrappers.exchange.batchFillOrKillOrders(
+      signedOrders,
+      takerAssetFillAmounts,
+      signatures
+    );
+  }
+
+  getCancelOrderTx(order) {
+    return this.contractWrappers.exchange.cancelOrder(order);
+  }
+
+  createOrder(makerAddress, sizeTh, pricePerTh) {
+    sizeTh = new BigNumber(sizeTh);
+    pricePerTh = new BigNumber(pricePerTh);
+    const expirationTime = new BigNumber(
+      Math.round(Date.now() / 1000) + 10 * 24 * 60 * 60
+    ); // 10 days
+    const exchangeAddress = this.contractWrappers.exchange.address;
+
+    // Encode the price of the sale into the maker asset data feed. This is used to pass the price to the
+    // honey lemon market contract proxy to enrich the sale event to make retrieval easier on the front end.
+    const makerAddetDataIncludingPrice = assetDataUtils.encodeERC20BridgeAssetData(
+      this.marketContractProxyAddress,
+      this.minterBridgeAddress,
+      web3.utils.utf8ToHex(pricePerTh.toString()) // this is the sale price within the data feed for the minterbride
+    );
+    const order = {
+      makerAddress, // maker is the first address (miner)
+      takerAddress: NULL_ADDRESS, // taker is open and can be filled by anyone (when an investor comes along)
+      makerAssetAmount: sizeTh, // The maker asset amount
+      takerAssetAmount: sizeTh.multipliedBy(pricePerTh), // The taker asset amount
+      expirationTimeSeconds: new BigNumber(expirationTime), // Time when this order expires
+      makerFee: new BigNumber(0), // 0 maker fees
+      takerFee: new BigNumber(0), // 0 taker fees
+      feeRecipientAddress: NULL_ADDRESS, // No fee recipient
+      senderAddress: NULL_ADDRESS, // Sender address is open and can be submitted by anyone
+      salt: generatePseudoRandomSalt(), // Random value to provide uniqueness
+      makerAssetData: makerAddetDataIncludingPrice,
```

Comment:
@andrei-anisimov what do you think about encoding this custom markerAssetData file so that the price can be emmited within the fill event? this is to make the price retrieval easier on the front end as all information is now encoded within the transaction.